shelf/bucket_update/search_updater.py

```diff
@@ -57,9 +58,46 @@ def run(self):
             self.update_manager.bulk_update(bulk_update)
             all_id_list = all_id_list + bulk_update.keys()
 
+        self.remove_unlisted_documents_per_bucket(all_id_list)
+
+    def prune(self):
+        """
+            Loops through a list of artifact paths, compares them against
+            the search index, and removes any indexes that no longer exist.
+        """
+        path_list = []
+
+        try:
+            path_list.extend(self.load_path_list())
+        except BucketNotFoundError:
+            self.logger.info("Bucket wasn't found. Leaving path_list empty.")
+
+        gc.collect()
+        self.logger.info("Starting to collect {0} artifact's Elasticsearch ids.".format(len(path_list)))
+        all_id_list = []
+
+        for chunk_list in self._chunk(path_list):
+            for path in chunk_list:
```

Comment:
I don't see a nice way of making this happen. Wouldn't I just have to iterate over the items again for both the update and prune functions as they do different things?